components/fxa-client/src/push.rs

```diff
@@ -0,0 +1,139 @@
+                let is_local_account = match &self.state.last_seen_profile {
+                    None => false,
+                    Some(profile) => profile.response.uid == account_uid,
+                };
+                Ok(if is_local_account {
+                    vec![AccountEvent::AccountDestroyed]
+                } else {
+                    vec![]
+                })
+            }
+            PushPayload::PasswordChanged | PushPayload::PasswordReset => {
+                let status = self.check_authorization_status()?;
+                Ok(if !status.active {
+                    vec![AccountEvent::AccountAuthStateChanged]
+                } else {
+                    vec![]
+                })
+            }
+            PushPayload::Unknown => {
+                log::info!("Unknown Push command.");
+                Ok(vec![])
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "command", content = "data")]
+pub enum PushPayload {
+    #[serde(rename = "fxaccounts:command_received")]
+    CommandReceived(CommandReceivedPushPayload),
+    #[serde(rename = "fxaccounts:profile_updated")]
+    ProfileUpdated,
+    #[serde(rename = "fxaccounts:device_connected")]
+    DeviceConnected(DeviceConnectedPushPayload),
+    #[serde(rename = "fxaccounts:device_disconnected")]
+    DeviceDisconnected(DeviceDisconnectedPushPayload),
+    #[serde(rename = "fxaccounts:password_changed")]
+    PasswordChanged,
+    #[serde(rename = "fxaccounts:password_reset")]
+    PasswordReset,
+    #[serde(rename = "fxaccounts:account_destroyed")]
+    AccountDestroyed(AccountDestroyedPushPayload),
+    #[serde(other)]
+    Unknown,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CommandReceivedPushPayload {
+    command: String,
+    index: u64,
+    sender: String,
+    url: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct DeviceConnectedPushPayload {
+    #[serde(rename = "deviceName")]
+    device_name: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct DeviceDisconnectedPushPayload {
+    #[serde(rename = "id")]
+    device_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct AccountDestroyedPushPayload {
+    #[serde(rename = "uid")]
+    account_uid: String,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_deserialize_send_tab_command() {
+        let json = "{\"version\":1,\"command\":\"fxaccounts:command_received\",\"data\":{\"command\":\"send-tab-recv\",\"index\":1,\"sender\":\"bobo\",\"url\":\"https://mozilla.org\"}}";
+        let _: PushPayload = serde_json::from_str(&json).unwrap();
+    }
+
+    #[test]
+    fn test_handle_push_message_unknown_command() {
+        let mut fxa =
+            FirefoxAccount::with_config(crate::Config::stable_dev("12345678", "https://foo.bar"));
+        let json = "{\"version\":1,\"command\":\"huh\"}";
+        let events = fxa.handle_push_message(json).unwrap();
+        assert!(events.is_empty());
+    }
```

Comment:
This PR adds some extra cases that are probably worth adding explicit tests for:

* We clear local profile data in response to profile-update.
* We clear local auth tokens when our own device is disconnected, but not when other devices are disconnected.